### ISSUE TITLE
Timeout error after ordering makes the algorithm place the same order twice

### ISSUE THREAD

**Issue by sam31415** [enigmampc/catalyst#317]
Dear Catalyst Maintainers,

Before I tell you about my issue, let me describe my environment:

# Environment

* Operating System: Linux trading-bot 4.13.0-39-generic #44-Ubuntu SMP Thu Apr 5 14:25:01 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
* Python Version: Python 3.6.4 :: Anaconda, Inc.
* Python Bitness: 64
* How did you install Catalyst: conda
* Python packages: 
aiodns                    1.1.1                     <pip>                      
aiohttp                   3.1.3                     <pip>                      
alembic                   0.9.9                     <pip>                      
async-timeout             2.0.1                     <pip>                      
attrdict                  2.0.0                     <pip>                      
attrs                     17.4.0                    <pip>                      
bcolz                     0.12.1                    <pip>                      
bcolz                     1.1.2            py36h00f5784_0                      
bokeh                     0.12.14                  py36_0                      
boto3                     1.7.4                     <pip>                      
botocore                  1.10.4                    <pip>                      
Bottleneck                1.2.1                     <pip>                      
bzip2                     1.0.6                h9a117a8_4                      
ca-certificates           2017.08.26           h1d4fec5_0                      
cchardet                  2.1.1                     <pip>                      
ccxt                      1.13.1                    <pip> 
certifi                   2018.1.18                py36_0                                                                                              
certifi                   2018.4.16                 <pip>                      
chardet                   3.0.4                     <pip>                      
click                     6.7              py36h5253387_0                      
cloudpickle               0.5.2                    py36_1                      
configparser              3.5.0                     <pip>                      
contextlib2               0.5.5                     <pip>                      
cycler                    0.10.0                    <pip>                      
cyordereddict             1.0.0                     <pip>                      
Cython                    0.28.2                    <pip>                      
cython                    0.27.3           py36h1860423_0                      
cytoolz                   0.9.0.1          py36h14c3975_0                      
dask                      0.17.2                   py36_0                      
dask-core                 0.17.2                   py36_0                      
decorator                 4.3.0                     <pip>                      
distributed               1.21.4                   py36_0                      
docutils                  0.14                      <pip>                      
empyrical                 0.2.1                     <pip>                      
enigma-catalyst           0.5.9                     <pip>                      
eth-abi                   1.0.0                     <pip>                      
eth-account               0.1.0a2                   <pip>                      
eth-hash                  0.1.2                     <pip>                      
eth-keyfile               0.5.1                     <pip>                      
eth-keys                  0.2.0b3                   <pip>                      
eth-rlp                   0.1.0                     <pip>                      
eth-utils                 1.0.2                     <pip>                      
hdf5                      1.10.1               h9caa474_1                      
heapdict                  1.0.0                    py36_2                      
hexbytes                  0.1.0                     <pip>                      
idna                      2.6                       <pip>                      
idna-ssl                  1.0.1                     <pip>                      
intel-openmp              2018.0.0                      8                      
intervaltree              2.1.0                     <pip>                      
jinja2                    2.10             py36ha16c418_0                      
jmespath                  0.9.3                     <pip>                      
kiwisolver                1.0.1                     <pip>                      
libedit                   3.1                  heed3624_0                      
libffi                    3.2.1                hd88cf55_4
libgcc-ng                 7.2.0                hdf63c60_3                                                                                              [38/577]
libgfortran-ng            7.2.0                hdf63c60_3                      
libstdcxx-ng              7.2.0                hdf63c60_3                      
locket                    0.2.0            py36h787c0ad_1                      
Logbook                   1.3.3                     <pip>                      
lru-dict                  1.1.6                     <pip>                      
lxml                      4.2.1                     <pip>                      
lzo                       2.10                 h49e0be7_2                      
Mako                      1.0.7                     <pip>                      
markupsafe                1.0              py36hd9260cd_1                      
matplotlib                2.2.2                     <pip>                      
mkl                       2018.0.2                      1                      
mpld3                     0.3                       <pip>                      
msgpack-python            0.5.5            py36h6bb024c_0                      
multidict                 4.2.0                     <pip>                      
multipledispatch          0.5.0                     <pip>                      
ncurses                   6.0                  h9df7e31_2                      
networkx                  2.1                       <pip>                      
numexpr                   2.6.4            py36hc4a3f9a_0                      
numpy                     1.14.2                    <pip>                      
numpy                     1.14.2           py36hdbf6ddf_0                      
openssl                   1.0.2n               hb7f436b_0                      
packaging                 17.1                     py36_0                      
pandas                    0.22.0           py36hf484d3e_0                      
pandas                    0.19.2                    <pip>                      
pandas-datareader         0.6.0                     <pip>                      
partd                     0.3.8            py36h36fd896_0                      
patsy                     0.5.0                     <pip>                      
patsy                     0.5.0                    py36_0                      
pip                       10.0.0                    <pip>                      
pip                       9.0.1                    py36_5                      
psutil                    5.4.3            py36h14c3975_0                      
pycares                   2.3.0                     <pip>                      
pycryptodome              3.6.1                     <pip>                      
pyparsing                 2.2.0            py36hee85983_1                      
pytables                  3.4.2            py36h3b5282a_2                      
python                    3.6.4                hc3d631a_3                      
python-dateutil           2.7.2                     <pip>
python-dateutil           2.7.0                    py36_0  
python-editor             1.0.3                     <pip>
pytz                      2018.3                   py36_0  
pytz                      2018.4                    <pip>
pyyaml                    3.12             py36hafb9ca4_1  
readline                  7.0                  ha6073c6_4  
redo                      1.6                       <pip>
requests                  2.18.4                    <pip>
requests-file             1.4.3                     <pip>
requests-ftp              0.3.1                     <pip>
requests-toolbelt         0.8.0                     <pip>
rlp                       0.6.0                     <pip>
s3transfer                0.1.13                    <pip>
scipy                     1.0.0            py36hbf646e7_0  
scipy                     1.0.1                     <pip>
setuptools                38.5.1                   py36_0  
setuptools                39.0.1                    <pip>
six                       1.10.0                    <pip>
six                       1.11.0           py36h372c433_1  
sortedcontainers          1.5.9                    py36_0  
SQLAlchemy                1.2.6                     <pip>
sqlite                    3.22.0               h1bed415_0  
statsmodels               0.8.0                     <pip>
statsmodels               0.8.0            py36h8533d0b_0  
tabulate                  0.8.2                     <pip>
tblib                     1.3.2            py36h34cf8b6_0  
tk                        8.6.7                hc745277_3  
toolz                     0.9.0                    py36_0  
tornado                   5.0                      py36_0  
urllib3                   1.22                      <pip>
web3                      4.1.0                     <pip>
websockets                4.0.1                     <pip>
wheel                     0.30.0           py36hfd4bba0_1  
wrapt                     1.10.11                   <pip>
xz                        5.2.3                h55aa19d_2  
yaml                      0.1.7                had09818_2  
yarl                      1.1.0                     <pip>
zict                      0.1.3            py36h3a3bf81_0  
zlib                      1.2.11               ha838bed_2

Now that you know a little about me, let me tell you about the issue I am
having:

# Description of Issue

I believe this issue is very similar to #315 . My algorithm trading on Binance crashed while trying to sell a position in zrx_btc, using effectively a market order (actually a limit order with price 0.9*current_price). Here is the log (my algo uses UTC time):

```
[2018-04-25 02:00:26.320708] INFO: CCXT: order amount adjusted by binance from 14.0 to 14.0
[2018-04-25 02:00:36.597066] WARNING: CCXT: unable to create order binance / IOST/BTC: binance POST https://api.binance.com/api/v3/order HTTPSConnectionPool(host='api.binance.com', port=443): Read timed out. (read timeout=10)
[2018-04-25 02:00:36.597524] WARNING: exchange_blotter: Ordering again. [2018-04-25 02:00:47.757379] WARNING: CCXT: unable to create order binance / IOST/BTC: binance {"code":-2010,"msg":"Account has insufficient balance for requested action."} [2018-04-25 02:00:47.757798] WARNING: exchange_blotter: Ordering again.[2018-04-25 02:00:57.717347] WARNING: CCXT: unable to create order binance / IOST/BTC: binance {"code":-2010,"msg":"Account has insufficient balance for requested action."} [2018-04-25 02:00:57.717717] WARNING: exchange_blotter: Ordering again. [2018-04-25 02:01:08.757048] WARNING: CCXT: unable to create order binance / IOST/BTC: binance {"code":-2010,"msg":"Account has insufficient balance for requested action."} [2018-04-25 02:01:08.757453] WARNING: exchange_blotter: Ordering again.[2018-04-25 02:01:26.270103] WARNING: CCXT: unable to create order binance / IOST/BTC: binance {"code":-2010,"msg":"Account has insufficient balance for requested action."}
```

Looking at my Binance account, the order was filled at 02:00 UTC (which appears as 04:00 because Binance is using my local timezone):
`2018-04-25 04:00:26    0.0001194414    0.00052997 BNB   0.00167216 BTC`
As far as I can tell, the order was filled against a single limit order, so this issue is different from #309.

So to summarize: It looks like the order was filled right when it was created, but that a timeout error prevented CCXT to realize that it went through. 

Maybe before trying to reorder, it would be good do add some logic that (for instance) checks whether the balance on the exchange changed, which would indicate a filled order. Also, insufficient balance exception should probably be catched (with a warning) instead of them making crash the algorithm. 

Here is how you can reproduce this issue on your machine:

## Reproduction Steps

1. This error is probably hard to reproduce as it depends on having a timeout at the right (wrong?) moment. Let algos run for a while and wait for timeouts...
...

Sincerely,
Samuel

**Comment by sam31415** [comment]
Same problem on Poloniex.
```
[2018-04-29 01:30:20.981939] WARNING: CCXT: unable to create order poloniex / BCH/BTC: poloniex POST https://poloniex.com/tradingApi HTTPSConnectionPoo[0/1847]
poloniex.com', port=443): Read timed out. (read timeout=10)                                                                                                    
[2018-04-29 01:30:20.982661] WARNING: exchange_blotter: Ordering again.                                                                                        
[2018-04-29 01:31:00.195456] INFO: exchange_blotter: filled order 76154149809 / bch_btc: 0.01727668, avg price: 0.155                                          
Traceback (most recent call last):                                                                                                                             
[...]                    
catalyst.exchange.exchange_errors.NotEnoughCashError: Total btc amount on poloniex is lower than the cash reserved for this algo: 0.01591353 < 0.01658084588291
68. While trades can be made on the exchange accounts outside of the algo, exchange must have enough free btc to cover the algo cash.
```
If I check my Poloniex account, I see that two identical orders went through (at 10 sec interval):
```
  | BCH/BTC | Buy | Exchange | 0.15500000 | 0.01727668 | 0.00004319 BCH (0.25%) | 0.00267788 BTC | 2018-04-29 01:30:33 
  | BCH/BTC | Buy | Exchange | 0.15500000 | 0.01727668 | 0.00004319 BCH (0.25%) | 0.00267788 BTC | 2018-04-29 01:30:23
```

**Comment by lenak25** [comment]
Hi @sam31415 ,

Thanks for catching this!
We were able to reproduce the timeout issue when running with a debugger (which was not the case in your scenario: It happened spontaneously after some time?) and saw the problem in the retry mechanism.
We are working to solve it.

Thanks,
Lena

**Comment by sam31415** [comment]
Hi @lenak25,

No, it happened spontaneously. What is this debugger?

Best,

Samuel

**Comment by lenak25** [comment]
Hi @sam31415 , I meant running with an IDE and using its debugger (Pycharm for example).
In any case, like you have mentioned it is the same issue as reported at #315, so we are closing this and will continue the discussion there.
Thanks for reporting!
